nova/compute/manager.py

```
@@ -5851,6 +5859,176 @@ def _complete_volume_attachments(self, context, bdms):
+                    device_type='disk', delete_on_termination=True,
+                    boot_index=-1, volume_size=ephemeral_size)
+                bdm.create()
+                reassing_device_names = True
+            elif len(ephemerals) == 1 and ephemeral_size == 0:
+                LOG.debug('Deleting ephemeral bdm.', instance=instance)
+                ephemerals[0].destroy()
+                reassing_device_names = True
+            elif len(ephemerals) == 1:
+                LOG.debug(
+                    f'Resizing ephemeral bdm: new_size={ephemeral_size}, '
+                    f'old_size={ephemerals[0].volume_size}.',
+                    instance=instance,
+                )
+                ephemerals[0].volume_size = ephemeral_size
+                ephemerals[0].save()
+
+        # Resize swap
+        if swap_size is not None:
+            if swap['swap_size'] == 0 and swap_size > 0:
+                LOG.debug(
+                    f'Creating swap bdm: size={swap_size}.',
+                    instance=instance,
+                )
+                bdm = objects.BlockDeviceMapping(
+                    context=context, instance_uuid=instance.uuid,
+                    source_type='blank', destination_type='local',
+                    device_type='disk', guest_format='swap',
+                    boot_index=-1, volume_size=swap_size)
+                bdm.create()
+                reassing_device_names = True
+            elif swap['swap_size'] > 0 and swap_size == 0:
+                LOG.debug('Deleting swap bdm.', instance=instance)
+                swap.destroy()
+                reassing_device_names = True
+            elif swap['swap_size'] != swap_size:
+                LOG.debug(
+                    f'Resizing swap bdm: old_size={swap.volume_size} '
+                    f'new_size={swap_size}',
+                    instance=instance,
+                )
+                swap.volume_size = swap_size
```

Same here: If you don't expect to shrink here, I would throw an error.